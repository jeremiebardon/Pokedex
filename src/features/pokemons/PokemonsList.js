import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Layout, Col, Row } from 'antd';

// Slice
import { fetchPokemons, resetState } from './PokemonsSlice';

// Components
import Loader from '../../components/Loader/Loader';
import PokemonItem from '../../components/PokemonItem/PokemonItem';
import PaginationBar from '../../components/Pagination/Pagination';

// Component Library
const { Header, Content, Footer } = Layout;

function PokemonsList() {
  const dispatch = useDispatch();
  const {
    pokemons,
    paginationPage,
    pokemonPerPage,
    count,
    loading,
  } = useSelector((state) => state.pokemons);
  const [paginateBy, setPaginateBy] = useState(pokemonPerPage || 50);
  const [currentPage, setCurrentPage] = useState(paginationPage || 1);

  useEffect(() => {
    dispatch(fetchPokemons({ paginateBy, currentPage }));
    return () => {
      dispatch(resetState());
    };
  }, [dispatch, currentPage, paginateBy]);

  function renderPokemonItems(pokemonsList) {
    return pokemonsList.map(
      (pokemon) => (
        <Col
          className="gutter-row"
          xs={12}
          sm={8}
          md={6}
          lg={4}
          key={pokemon.id}
        >
          <PokemonItem
            key={pokemon.id}
            id={pokemon.id}
            image={pokemon.sprites.front_default}
            name={pokemon.name}
            types={pokemon.types}
          />
        </Col>
      ),
    );
  }

  function handlePagination(page, itemOnPage) {
    setCurrentPage(page);
    setPaginateBy(itemOnPage);
  }

  return (
    <>
      { loading && <Loader /> }

      <Header style={{ textAlign: 'center', color: '#fff', fontWeight: '800' }}>
        Poke API
      </Header>

      <Content style={{ padding: '16px', paddingBottom: '80px' }}>
        <Row gutter={16}>
          { pokemons && !loading && (
            renderPokemonItems(pokemons)
          )}
        </Row>
      </Content>
      <Footer style={{ position: 'fixed', bottom: '0', width: '100%' }}>
        <PaginationBar
          pageLoad={loading}
          itemsCount={count}
          itemsPerPage={paginateBy}
          handlePagination={handlePagination}
          currentPage={currentPage}
        />
      </Footer>
    </>
  );
}

export default PokemonsList;
