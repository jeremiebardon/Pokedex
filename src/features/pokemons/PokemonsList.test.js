import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { test, expect } from 'jest-without-globals';
import '../../../__mocks__/matchMedia';
import * as redux from 'react-redux';

import PokemonsList from './PokemonsList';

test('Render PokemonsList', () => {
  const useDispatchSpy = jest.spyOn(redux, 'useDispatch');
  const useSelectorSpy = jest.spyOn(redux, 'useSelector');
  const mockDispatchFn = jest.fn();
  const mockSelectorFn = jest.fn();
  useDispatchSpy.mockReturnValue(mockDispatchFn);
  useSelectorSpy.mockReturnValue(mockSelectorFn);

  const component = render(<PokemonsList />, { wrapper: MemoryRouter });
  expect(component).toMatchSnapshot();
});
