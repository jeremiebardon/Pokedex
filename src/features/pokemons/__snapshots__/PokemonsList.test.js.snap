// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render PokemonsList 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="ant-layout-header"
        style="text-align: center; color: rgb(255, 255, 255); font-weight: 800;"
      >
        Poke API
      </header>
      <main
        class="ant-layout-content"
        style="padding: 16px 16px 80px 16px;"
      >
        <div
          class="ant-row"
          style="margin-left: -8px; margin-right: -8px;"
        />
      </main>
      <footer
        class="ant-layout-footer"
        style="position: fixed; bottom: 0px; width: 100%;"
      />
    </div>
  </body>,
  "container": <div>
    <header
      class="ant-layout-header"
      style="text-align: center; color: rgb(255, 255, 255); font-weight: 800;"
    >
      Poke API
    </header>
    <main
      class="ant-layout-content"
      style="padding: 16px 16px 80px 16px;"
    >
      <div
        class="ant-row"
        style="margin-left: -8px; margin-right: -8px;"
      />
    </main>
    <footer
      class="ant-layout-footer"
      style="position: fixed; bottom: 0px; width: 100%;"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
